### PR TITLE
Add support for more than 3 cuda streams and custom synchronization between them

### PR DESCRIPTION
**Pull request by andrerenard**
Currently there are 3 cuda streams, two for copies, and one for compute.  This change allows for an arbitrary number of streams set by the `num_streams` option.   When using this option it becomes necessary to  provide a `cuda_stream` for each commandObject. 

To synchronize between streams, a new command object called `cudaSyncStream` can be used to place cuda wait events which depend on other streams. 

All configurations moving forward will need to use this object to sync between the kernel and copy streams.

### CODE REVIEW

**Comment by andrerenard** [comment]
@dstndstn I still need to do a few more correctness tests, but this seems to be passing the initial tests, and it's ready for review.

**Comment by dstndstn** [comment]
(no rush, just putting this here while I think of it --)
@andrerenard , I'd appreciate your comment on [0744ecb](https://github.com/kotekan/kotekan/pull/1076/commits/0744ecbaa0fc4c3903479b4b1e9e1ef646cb857e) -- I added an additional cudaSyncStream constructor that doesn't run the initialization steps whose arguments come from the config file, so I can do something different in the cudaSyncInput constructor.  Maybe you want to recommend a better way to do this.

**Review by dstndstn — APPROVED** [review]
This seems reasonable to me!

**Comment by dstndstn** [comment]
By the way, it looks like the one failing CI test, the parallel python tests, are consistently failing, not just intermittently from a floating-point issue, but related to setting memory policy:
```
Creating visBuffer named dumpvis_buf0 with 4 frames, frame size of 176 and metadata pool vis_pool on numa_node 0
Failed to set memory policy: Operation not permitted (1)
Failed to start kotekan with config file /tmp/tmpdd2cyatj, error message: Could not create the buffer: dumpvis_buf0
Exiting...
```

**Comment by dstndstn** [comment]
@andrerenard From my perspective, I think this is ready, but you should review my changes.  (I don't seem to be able to request a review from you, since it's your PR?)  Thanks!

File: lib/cuda/cudaBasebandBeamformer.cpp
Comment (andrerenard) [inline lib/cuda/cudaBasebandBeamformer.cpp]: I think we can actually remove the wait event on the pre_event.  Since the pre_event is now on the same stream streams, this doesn't provide any meaningful synchronization past what the stream ordering guarantees already.   Before this was used to sync streams, but that's now provided by `syncStream`.

File: lib/cuda/cudaSyncStream.cpp
Comment (andrerenard) [inline lib/cuda/cudaSyncStream.cpp:19]: This still needs a command type set for the internal profiling.  
```suggestion
    (void)called_by_subclasser;
    command_type = gpuCommandType::BARRIER;
```

File: lib/cuda/cudaSyncStream.hpp
Comment (andrerenard) [inline lib/cuda/cudaSyncStream.hpp]: I cannot really think of a better way to do this.  However, I might add a little more to this comment to explain what's different.  e.g. doesn't get streams from the configuration file.

**Review by andrerenard — COMMENTED** [review]
Thanks for the changes!   It looks much cleaner now for the default case.   I cannot "request changes" on my own PR, but there are a few small changes in this comment we should do before merging.

File: lib/cuda/cudaBasebandBeamformer.cpp
Comment (dstndstn) [reply lib/cuda/cudaBasebandBeamformer.cpp]: Does this go for all cudaCommand compute subclasses?  Eg this same snippet also appears in `cudaCorrelatorAstron.cpp`, `cudaCorrelator.cpp` and `cudaOutputDataZero.cpp`.

File: lib/cuda/cudaSyncStream.cpp
Comment (dstndstn) [reply lib/cuda/cudaSyncStream.cpp:19]: Done!

File: lib/cuda/cudaSyncStream.hpp
Comment (dstndstn) [reply lib/cuda/cudaSyncStream.hpp]: done

File: lib/cuda/cudaBasebandBeamformer.cpp
Comment (andrerenard) [reply lib/cuda/cudaBasebandBeamformer.cpp]: Yes I think we can remove this from all subclasses (other than the sync of course).

Comment (dstndstn) [reply lib/cuda/cudaBasebandBeamformer.cpp]: Right, I see what you mean now, these were just waiting on the previous event in their stream, which is... already how a stream works :)  Done!

**Comment by dstndstn** [comment]
Just the usual CI test failing (python test failing to set memory policy)!  Good to merge?

**Comment by andrerenard** [comment]
Hmm, that's a new failure mode... I'm going to rerun it and see if it's consistent.

**Comment by dstndstn** [comment]
FWIW I also see it in https://github.com/kotekan/kotekan/pull/1078

**Comment by andrerenard** [comment]
Ah, looks like I broke that with another PR, but didn't catch it because it was also failing with the unrelated float issue.  It doesn't impact this PR though, so I'll merge this in, and then fix that separately.

**Pull request merged**
